pyof/v0x04/common/flow_match.py

```diff
@@ -182,17 +182,101 @@ class VlanId(IntEnum):
 
 # Classes
 
+class OxmTLV(GenericStruct):
+    """Oxm (OpenFlow Extensible Match) TLV."""
+
+    oxm_class = UBInt16(enum_ref=OxmClass)
+    oxm_field_and_mask = UBInt8()
+    oxm_length = UBInt8()
+    oxm_value = BinaryData()
+
+    def __init__(self, oxm_class=None, oxm_field=None,
+                 oxm_hasmask=False, oxm_value=None):
+        """Inialize OXM TLV struct.
+
+        Args:
+            oxm_class (OxmClass): Match class: member class or reserved class
+            oxm_field (OxmMatchFields, OxmOfbMatchField): Match field within
+                the class
+            oxm_hasmask (bool): Set if OXM include a bitmask in payload
+            oxm_value (bytes): OXM Payload
+        """
+        super().__init__()
+        self.oxm_class = oxm_class
+        self.oxm_field_and_mask = None
+        self.oxm_length = None
+        self.oxm_value = oxm_value
+        # Attributes that are not packed
+        self.oxm_field = oxm_field
+        self.oxm_hasmask = oxm_hasmask
+
+    def unpack(self, buff, offset=0):
+        """Unpack the buffer into a OxmTLV.
 
-class OxmHeader(GenericStruct):
+        Args:
+            buff (bytes): The binary data to be unpacked.
+            offset (int): If we need to shift the beginning of the data.
+        """
+        super().unpack(buff, offset)
+        # Recover field and hasmask from field_and_hasmask
+        oxm_field_value = self.oxm_field_and_mask >> 1
+        try:
+            self.oxm_field = OxmOfbMatchField(oxm_field_value)
+        except ValueError:
+            raise UnpackException(f'Invalid oxm_field: "{self.oxm_field}"')
+        self.oxm_hasmask = self.oxm_field_and_mask & 1 == 1
+
+        # Unpack oxm_value that has oxm_length bytes
+        start = offset + 4  # 4 bytes: class, field_and_mask and length
+        end = start + self.oxm_length
+        self.oxm_value = buff[start:end]
+
+    def _update_length(self):
+        payload = type(self).oxm_value.pack(self.oxm_value)
+        self.oxm_length = len(payload)
+
+    def pack(self, value=None):
+        """Join oxm_hasmask bit and 7-bit oxm_field."""
+        if value is not None:
+            return value.pack()
+
+        # Make sure oxm_field is valid
+        if isinstance(self.oxm_field, OxmOfbMatchField):
+            field_enum = self.oxm_field
+        else:
+            try:
+                field_enum = OxmOfbMatchField(self.oxm_field)
+            except ValueError:
+                raise PackException(f'Invalid oxm_field: "{self.oxm_field}"')
+
+        # Set oxm_field_and_mask instance attribute
+        # 1. Move self.oxm_field one bit to the left and truncate it to 7 bits
+        oxm_field = field_enum.value << 1 & 127
```

Comment:
This is still incorrect. You shouldn't bitAND it with 127. see https://github.com/kytos/python-openflow/issues/348#issuecomment-318068024